### ISSUE TITLE
Add a SEPA QR payment code in the standard Direct Bank Transfer

### ISSUE THREAD

**Issue by Syprer** [woocommerce/woocommerce#27661]
**Is your feature request related to a problem? Please describe.**
When customers need to type over bank information and memo field info they could make writing mistake

**Describe the solution you'd like**
In europe, the EPC has made a SEPA-payment QR code image live rendered from their servers ( <www.epc-qr.eu> )

**Describe alternatives you've considered**
I've added this php line in the email confirmation, but this should only be printed when the payment method is "direct bank transfer"
echo " <img src='https://epc-qr.eu/?bname=Vendor%20Name&iban=".$vendor->iban."&euro=". $order->get_total() ."&info=Order%20Number%20". $order->get_id() ."%20&bic=".$vendor->aba_routing_number."' />" ;

**Additional context**
This barcode is accepted by many bank apps from Europe and the payments are handled inside the app by SEPA rules.

example donation of 1 euro to the Red Cross
<img src='https://epc-qr.eu/?bname=Red%20Cross%20Belgium&iban=BE72000000001616&euro=1&info=Donation%20from%20Woocommerce&bic=BPOTBEB1' />"

**Comment by juliaamosova** [comment]
Hi @Syprer,

Thank you for taking the time to share this idea, we really appreciate your help. 

While this is clearly a great improvement/idea we won’t be able to tackle it in the upcoming weeks. We’re going to add it to our backlog where it will be considered for future releases.

**Comment by sebastijanz** [comment]
Hi there, this would be really great improvement!
I hope that can be developed soon.

**Comment by tinof** [comment]
With Transferwise, a Woocommerce store owner could use SEPA QR code for easy payment  and get instant verification when Transferwise API has confirmed the payment. This way the user never has to login to their bank (no man in the middle attacks) or give bank account details (SEPA Direct Payment)?

**Comment by sebastijanz** [comment]
Anything new on this issue?

**Comment by mxa** [comment]
We need this. Its essential.

It's the girocode, or more officially called the [European Payments Council Quick Response Code](https://en.wikipedia.org/wiki/EPC_QR_code). There is a guideline [whitepaper on the website of the European Payments Council](https://www.europeanpaymentscouncil.eu/document-library/guidance-documents/quick-response-code-guidelines-enable-data-capture-initiation)
For now there is an Open Source Wordpress plugin which should give you shortcode access, but it seems broken., I can't use this shortcode without an error.
https://wordpress.org/plugins/sepa-girocode/

**Comment by Coernel82** [comment]
**Update: Had a look at the sourcode - potentially malicous and GDPR-breach! Passes on the data to an external webserver.**

GiroCode by Minerva Kulturdienstleistungen UG (haftungsbeschränkt) (https://documentid.net/) does work. By using a shortcode such as [girocode beneficiary="Deutsches Rotes Kreuz e. V." iban="DE63370205000005023307" amount="10.00" purpose="Internationale Soforthilfe"] (I googled the iban, it is really the red cross) the QR-Code gets created. So the first step is done. All what is missing is that for iban, amount and purpose there should be variables! Place this is the checkout.php or whatever the file is called and done!
However I do not know how to program this so it would be great if someone solves this and creates a pull request!

**Comment by mxa** [comment]
I think this functionality should be part of the internal BACS Payment method.

**Comment by Coernel82** [comment]
> 
> 
> I think this functionality should be part of the internal BACS Payment method.

Yes! I agree - that is where it belongs.

**Comment by Coernel82** [comment]
**Update: Had a look at the sourcode - potentially malicous and GDPR-breach! Passes on the data to an external webserver.**

Now after some hours of trial and error some ideas for the solution:
In conjunction with the [GiroCode Plugin](https://wordpress.org/plugins/girocode/) being activated:
 
On the page **before** the `thankyou.php` this works:

```
<?php	
$testiban = "DE63370205000005023307";
$amount_qr_code= WC()->cart->total; 
echo do_shortcode('[girocode beneficiary="Deutsches Rotes Kreuz e.V." IBAN="'.$testiban.'" amount="'.$amount_qr_code.'" purpose="Internationale Soforthilfe" size="200"]');
?>
```

On the thankyou.php page however the QR-Code stays empty - I guess `WC()->cart->total; ` is not working there.
Also for **purpose** there should not be a hardcoded purpose but the order number. Does it work this way? `$order->get_id();`
This does not work on the page before the thank you page - I guess mainly because there is no order ID existant yet.

Can someone help with the next steps? @Syprer - you seem to have basic programming skills.

**Comment by Coernel82** [comment]
Just had a look at the sourcecode of the GiroCode Plugin - it just passes on parameters to the developers website and loads a png from there. So this is not GDPR-Compliant! However there are public libraries: https://github.com/topics/girocode?l=php

**Comment by mxa** [comment]
> Just had a look at the sourcecode of the GiroCode Plugin - it just passes on parameters to the developers website and loads a png from there. So this is not GDPR-Compliant! However there are public libraries: https://github.com/topics/girocode?l=php

Wow, this is potentially malicious.
We don't need a special _Girocode_ library, It's a standard QR-code in a special syntax which makes it a "Girocode".

**Comment by matthiaswagnerwy** [comment]
we would definitely need to include the generation of the code into core. all solutions i have met so far depend on some apis, which is not only problematic regarding gdpr. it is also a problem that we have delay and can not rely on them to be always online. also, best solution would be to generate svg images.

here is a MIT licensed javascript project for that: https://github.com/papnkukn/qrcode-svg

**Comment by Coernel82** [comment]
The php-File from 
fellwell5 /
bezahlcode 

on GitHub creates the QR-Code ~from the Google API. So useless as well. But I guess there are libaries already available.~ The developer updated his plugin upon my request so that it is usable without Google API.

**Comment by Coernel82** [comment]
> 
> 
> we would definitely need to include the generation of the code into core. all solutions i have met so far depend on some apis, which is not only problematic regarding gdpr. it is also a problem that we have delay and can not rely on them to be always online. also, best solution would be to generate svg images.
> 
> here is a MIT licensed javascript project for that: https://github.com/papnkukn/qrcode-svg

That indeed is the best option as there is absolutely no data sent somewhere. So really what has to be done is simply to create a string using Woocommerce variables being compliant with [https://en.wikipedia.org/wiki/EPC_QR_code](url)

As said before I am not a programmer however I will have a look at this! Creating a string and passing the variables onto the javascript code could not be to difficult, right?

Can someone help how to access the necessary variables from woocommerce for IBAN, recipient reference, amount and order number?

**Comment by matthiaswagnerwy** [comment]
since woocommerce is mostly server-based, we need to find a php algorithm or library (that is fine to include in woocommerce) that converts the string to a svg source. javascript won't be good here, but maybe it could be a hint on how to implement the algorithm. that's why i linked it here ;)

**Comment by mxa** [comment]
http://phpqrcode.sourceforge.net/

**Comment by Coernel82** [comment]
One step further because it already implements the SEPA-String and after a request the programmer included the qr-creation without google-api.
[https://github.com/fellwell5/bezahlcode](url)

**Comment by Coernel82** [comment]
I know that this is basic programing skill however if someone tells me how to get the **IBAN** and **BIC** out of [this array](https://woocommerce.github.io/code-reference/files/woocommerce-includes-gateways-bacs-class-wc-gateway-bacs.html#source-view.307) beginning in Line 307 as **seperate variables** I can provide you with a working solution. It may not be the proper way but it will work.

**Comment by designboxcom** [comment]
I am in the same boat here.  Many European Customers now need a Girocode to simplify the Wire transfers. However, it seems https://wordpress.org/plugins/girocode/ has been disabled. Does Woocommerce plans to include this function into the core anytime soon ?

**Comment by Coernel82** [comment]
@designboxcom Yes that plugin has been disabled due to my abuse report! Though this plugin creates valid girocodes by using a shortcode it sends data (to be clear: the payment data - IBAN, BIC, amount etc.) to another server. This was clear in line 22 in the sourcecode. I 1 star rated the plugin noting that this is potentially malicous and clearly not GDPR-compliant. The issuer of the plugin just deleted my rating saying I was asking for support (which I was not). I was saying: do not use, might collect banking data.
So after this highly suspicious behavior I had the plugin being deactivated.

Furthermore the plugin didn't solve the problem as it was only possible to "hardcode" the data including the amount in the shortcode.

Have a look at my post before yours - when someone solves this quite easy question I will provide you with a working solution and really easy step by step instructions how to get it work. It will have to be added manually by placing two or three files in one directory and adding some lines to one file calling that new function. It will not fulfill the Woocommerce programing codex etc. but it will work 100% automatically. I have tested it so far successfully and you could hardcode your IBAN and BIC into the sourcecode but I want this to be done using the array mentioned above.

**Comment by designboxcom** [comment]
@Coernel82 It totally makes sense. Thank you very much for your update. I'll be happy to try your solution as soon as someone will help to resolve your question.

**Comment by designboxcom** [comment]
@Coernel82 It seems the girocode plugin is back online. Still I will wait for your solution as I also prefer no third parties got involved in the generation process....

**Comment by Coernel82** [comment]
Short update: I have a real programer on it who will create a solid solution. It's a matter of some days and I will post it here!

**Comment by Coernel82** [comment]
Status update: The code is working already with one minor bug which hopefully will get solved the next days. The code will come as a plugin. No playing around with the functions.php just manually install and aktivate the plugin. You might need to hardcode your IBAN though.

On question: This thread is quite full - is it possible that someone deletes everything so that there will just be the link to the solution? Or do I open a new thread and someone closes this thread? @Syprer as you started this - which options do you have in this regard?

**Comment by vortexfoto** [comment]
@Coernel82 hello, did you manage to create plugin, im verry interested as its almost mandatory for us to use QR code in bank transfer. Thanks for reply

**Comment by Coernel82** [comment]
@vortexfoto Yes. And the one little bug was just a typo. Also there is no need to hardcode the IBAN. It is working perfectly: Install. Activate. Done. However I will have a look at the code to understand a bit then publish it on my github profile and write some proper documentation. Give me another one or two days!

**Comment by vortexfoto** [comment]
@Coernel82 thanks :)

**Comment by sebastijanz** [comment]
@Coernel82 Thats perfect. Cant wait to try it :) I'm waiting for some solution already for one year :)

@juliaamosova If you are still here, Will WOO implement this also? :)

**Comment by Coernel82** [comment]
As this thread is cluttered:

# Working SEPA-QR-Code plugin for Woocommerce
## discussion
https://github.com/woocommerce/woocommerce/issues/31358

## directly to the plugin
https://github.com/Coernel82/SEPA-QR-for-Woocommerce

Please tell me if it works and if the documentation is clear!

**Maybe someone can close this thread?**

**Comment by sebastijanz** [comment]
@Coernel82 I have tried to scan the qr code with my bank application, but code is not recognised :(

**Comment by Coernel82** [comment]
Which QR code? A sample code or a code generated with your Woocommerce?
Scan it with a normal QR-Code scanner, post the result (the text) here.

**Comment by Coernel82** [comment]
@Syprer @sebastijanz @designboxcom @vortexfoto @mxa @matthiaswagnerwy @tinof 
Do you have any feedback? I found the first tiny bug.
Maybe you can give some feedback and report bugs here: https://github.com/Coernel82/SEPA-QR-for-Woocommerce/issues

**Comment by Bumjigga101** [comment]
Hi there, 
Firstly, thanks for the effort put into developing this mod. I've been looking for something similar to this for a couple
of years on and off.. The intention was always just to do it myself when i got the time, but alas, here I am looking again.
I'd like to point out that, from this thread, there seems to be a lack of understanding of the GDPR which is self imposing unneeded restrictions on it's development. While i support the concept of self generating the QR codes as a more secure method to avoid mitm vulnerabilities, the GDPR isn't really relevant here.. No personal / identifiable information is passed at any point, to third parties or otherwise. The only information passed is the IBAN / BIC (which is public facing by default, and a deposit only payment method), payment amount and order number (if anyone other than the customer or vendor can identify an individual from this, you have bigger issues than this mod).
Secondly, I had previously found a girobank QR code generator that embeded a girobank link into the QR code (I do hope it wasn't the unjust report that saw it removed). But this, at the time, was a generic script and not woocommerce specific, so i didn't implement it. This embeded link generation, along with the visual QR code, as an option for customers, is something i was looking for. This link passes paramenters directly to adoptive banking apps, prepopulating payment details in app, much as  scanning the QR code does. The main reason for wanting this is.. We're looking at generating a QR code to be scanned by a mobile device, but as the majority of web traffic is now mobile, how do you scan a QR code being displayed on the scanning devices screen? (I'm sure theres device specific methods, copy paste etc.. but the point is making the process easier and less error prone for the customer) ;)
For now, I'll follow this as an 'almost, but not quite' option, in the hope that any development doesn't inhibit development of the concept overall. It's a shame there isn't already, but a properly developed and complete solution here would see thousands of willingly paying retailers throughout europe.

Sorry if any of this came accross as harsh, not the greatest with words, but i do genuinely appreciate and thank you for the efforts made ;)

**Comment by vortexfoto** [comment]
hello guys, could anyone please tell me how to show generated QR code in html code or in like elementor.
I am trying to include this QR code to PDF proforma invoice....

**Comment by mxa** [comment]
Why would you want that someone pays a proforma invoice?

**Comment by holle75** [comment]
Oh my God! i was trying the same thing for years now. Getting on programmers nervs around the world.
cornel82, i´m kissing your feet!

And yes, this should be in the core!

Just a silly question: why is msx-sepaqr.php calling vendor/bezahlcode.class.php where https://chart.apis.google.com is called?

**Comment by Coernel82** [comment]
I am happy to announce a milestone:
I released the plugin. Next step: Localization. Help wanted!
https://wordpress.org/plugins/mxp-sepa-qr-code-addon-for-woocommerce

**Comment by AlkoKod** [comment]
> I am happy to announce a milestone: I released the plugin. Next step: Localization. Help wanted! https://wordpress.org/plugins/mxp-sepa-qr-code-addon-for-woocommerce

Hello @Coernel82, it looks like your GitHub repo is archived so you no longer maintain this plugin?

**Comment by petermorlion** [comment]
@AlkoKod That's correct. As I was in discussion with him for image support and need it myself, I created [a fork](https://github.com/petermorlion/SEPA-QR-for-Woocommerce/tree/main) and am prepared to continue the work. Although in the future, I believe it should become obsolete entirely as SEPA QR codes should become part of WooCommerce.

**Comment by Coernel82** [comment]
Yes, I had to close it due to some reasons: Firstly my wife will have a baby in summer and also I have had an anxiety disorder for 7 months now. This is draining mental resources. As I am not a programmer and this was entirely a hobby project it was very time consuming. This time I do not have any more.
I will update the readme!

**Comment by petermorlion** [comment]
Congrats on the baby @Coernel82 ! Take your time and focus on mental health. For yourself and your family. Big thank you for your work on the plugin!

**Comment by Coernel82** [comment]
Thank you very much @petermorlion - we are looking forward to the baby. I will have a look at your fork from time to time and contribute!

**Comment by stgoos** [comment]
> Although in the future, I believe it should become obsolete entirely as SEPA QR codes should become part of WooCommerce.

I also would perfer to see this kind of functionality within the WooCommerce core as it will be useful for all (European based) stores which accept SEPA payments.

Thanks for the nice job @Coernel82 and @petermorlion!

**Comment by mxa** [comment]
I just discovered that there is no SEPA payment in woocommerce at all, there is a BACS payment option in the "offline payment" section, but BACS is just available in the UK and a few other countries, a small subset of the [SEPA countries](https://www.ecb.europa.eu/paym/integration/retail/sepa/html/index.en.html). In the Stripe or PayPal plugins there are SEPA Debit (Lastschrift) options, but that goes to their accounts and not directly to you.
